cpp/bench/sg/dbscan.cu

```diff
@@ -0,0 +1,116 @@
+/*
+ * Copyright (c) 2019, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <cuML.hpp>
+#include <dbscan/dbscan.hpp>
+#include <utility>
+#include "benchmark.cuh"
+
+namespace ML {
+namespace Bench {
+namespace dbscan {
+
+struct AlgoParams {
+  int min_pts;
+  double eps;
+  size_t max_bytes_per_batch;
+};
+
+struct Params {
+  DatasetParams data;
+  BlobsParams blobs;
+  AlgoParams dbscan;
+};
+
+template <typename D>
+class Dbscan : public BlobsFixture<D> {
+ public:
+  Dbscan(const std::string& name, const Params& p)
+    : BlobsFixture<D>(p.data, p.blobs), dParams(p.dbscan) {
+    this->SetName(name.c_str());
+  }
+
+ protected:
+  void runBenchmark(::benchmark::State& state) override {
+    if (!this->params.rowMajor) {
+      state.SkipWithError("Dbscan only supports row-major inputs");
+    }
+    auto& handle = *this->handle;
+    auto stream = handle.getStream();
+    for (auto _ : state) {
+      CudaEventTimer timer(handle, state, true, stream);
+      dbscanFit(handle, this->data.X, this->params.nrows, this->params.ncols,
+                D(dParams.eps), dParams.min_pts, labels,
+                dParams.max_bytes_per_batch);
```

Comment:
I kind of like @JohnZed's suggestion that we should probably be verifying if the output is correct while we are stress testing / benchmarking. I've had a couple different situations now where I was getting extremely good numbers until I found the algorithms either weren't converging (stopping prematurely because of a bug or bad hyperparam) or they were just wrong.

Reply:
Completely agree with your point. I have already filed issue #1083 for this exact purpose. Let's discuss more details in that issue.